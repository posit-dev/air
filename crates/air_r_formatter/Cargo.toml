
[package]
authors.workspace = true
categories.workspace = true
description = "Air R Formatter"
edition.workspace = true
homepage.workspace = true
keywords.workspace = true
license.workspace = true
name = "air_r_formatter"
publish = false
repository.workspace = true
rust-version.workspace = true
version = "0.0.0"

[dependencies]
air_r_syntax = { workspace = true }
biome_formatter = { workspace = true }
biome_rowan = { workspace = true }
itertools = { workspace = true }
settings = { workspace = true, features = ["biome", "serde"] }
tracing = { workspace = true }

[dev-dependencies]
air_formatter_test = { workspace = true }
air_r_parser = { workspace = true }
biome_parser = { workspace = true }
line_ending = { workspace = true }
tests_macros = { workspace = true }

[lints]
workspace = true
