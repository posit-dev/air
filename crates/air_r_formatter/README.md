To run tests you need to install [cargo insta](https://crates.io/crates/cargo-insta):

```sh
cargo install cargo-insta
```

Run and review tests with:

```sh
cargo insta test
cargo insta review
```

The `src/generated.rs` file is produced by running this at the workspace root:

```sh
just gen-formatter
```
