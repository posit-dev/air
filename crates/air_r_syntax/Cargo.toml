[package]
name = "air_r_syntax"
version = "0.0.0"
authors.workspace = true
edition.workspace = true
rust-version.workspace = true
description = "SyntaxKind and common rowan definitions for air_r_parser"
homepage.workspace = true
repository.workspace = true
license.workspace = true
keywords.workspace = true
categories.workspace = true
publish = false

[dependencies]
biome_rowan = { workspace = true, features = ["serde"] }
serde = { workspace = true, features = ["derive"] }

[lints]
workspace = true
