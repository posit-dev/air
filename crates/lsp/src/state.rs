use std::collections::HashMap;

use anyhow::anyhow;
use url::Url;

use crate::documents::Document;

#[derive(Clone, Default, Debug)]
/// The world state, i.e. all the inputs necessary for analysing or refactoring
/// code. This is a pure value. There is no interior mutability in this data
/// structure. It can be cloned and safely sent to other threads.
pub(crate) struct WorldState {
    /// Watched documents
    pub(crate) documents: HashMap<Url, Document>,

    /// The scopes for the console. This currently contains a list (outer `Vec`)
    /// of names (inner `Vec`) within the environments on the search path, starting
    /// from the global environment and ending with the base package. Eventually
    /// this might also be populated with the scope for the current environment
    /// in debug sessions (not implemented yet).
    ///
    /// This is currently one of the main sources of known symbols for
    /// diagnostics. In the future we should better delineate interactive
    /// contexts (e.g. the console, but scripts might also be treated as
    /// interactive, which could be a user setting) and non-interactive ones
    /// (e.g. a package). In non-interactive contexts, the lexical scopes
    /// examined for diagnostics should be fully determined by variable bindings
    /// and imports (code-first diagnostics).
    ///
    /// In the future this should probably become more complex with a list of
    /// either symbol names (as is now the case) or named environments, such as
    /// `pkg:ggplot2`. Storing named environments here will allow the LSP to
    /// retrieve the symbols in a pull fashion (the whole console scopes are
    /// currently pushed to the LSP), and cache the symbols with Salsa. The
    /// performance is not currently an issue but this could change once we do
    /// more analysis of symbols in the search path.
    #[allow(dead_code)]
    pub(crate) console_scopes: Vec<Vec<String>>,

    /// Currently installed packages
    #[allow(dead_code)]
    pub(crate) installed_packages: Vec<String>,
}

impl WorldState {
    pub(crate) fn get_document(&self, uri: &Url) -> Option<&Document> {
        self.documents.get(uri)
    }

    pub(crate) fn get_document_mut(&mut self, uri: &Url) -> Option<&mut Document> {
        self.documents.get_mut(uri)
    }

    pub(crate) fn get_document_or_error(&self, uri: &Url) -> anyhow::Result<&Document> {
        match self.get_document(uri) {
            Some(doc) => Ok(doc),
            None => Err(anyhow!("Can't find document for URI {uri}")),
        }
    }

    pub(crate) fn get_document_mut_or_error(&mut self, uri: &Url) -> anyhow::Result<&mut Document> {
        match self.get_document_mut(uri) {
            Some(doc) => Ok(doc),
            None => Err(anyhow!("Can't find document for URI {uri}")),
        }
    }

    pub(crate) fn workspace_uris(&self) -> Vec<Url> {
        self.documents.iter().map(|elt| elt.0.clone()).collect()
    }
}
