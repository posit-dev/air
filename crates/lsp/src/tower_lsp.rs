//
// tower_lsp.rs
//
// Copyright (C) 2022-2024 Posit Software, PBC. All rights reserved.
//
//

#![allow(deprecated)]

use strum::IntoStaticStr;

use tokio::io::{AsyncRead, AsyncWrite};
use tokio::sync::mpsc::unbounded_channel as tokio_unbounded_channel;
use tower_lsp::Client;
use tower_lsp::LanguageServer;
use tower_lsp::LspService;
use tower_lsp::jsonrpc::Result;
use tower_lsp::lsp_types::*;
use tower_lsp::{ClientSocket, jsonrpc};

use crate::handlers_ext::ViewFileParams;
use crate::main_loop::Event;
use crate::main_loop::GlobalState;
use crate::main_loop::TokioUnboundedSender;

// Based on https://stackoverflow.com/a/69324393/1725177
macro_rules! cast_response {
    ($target:expr, $pat:path) => {{
        match $target {
            Ok($pat(resp)) => Ok(resp),
            Err(err) => Err(new_jsonrpc_error(format!("{err:?}"))),
            _ => panic!("Unexpected variant while casting to {}", stringify!($pat)),
        }
    }};
}

#[allow(clippy::large_enum_variant)]
#[derive(Debug)]
pub(crate) enum LspMessage {
    Notification(LspNotification),
    Request(
        LspRequest,
        TokioUnboundedSender<anyhow::Result<LspResponse>>,
    ),
}

#[derive(Debug, IntoStaticStr)]
pub(crate) enum LspNotification {
    Initialized(InitializedParams),
    DidChangeWorkspaceFolders(DidChangeWorkspaceFoldersParams),
    DidChangeConfiguration(DidChangeConfigurationParams),
    DidChangeWatchedFiles(DidChangeWatchedFilesParams),
    DidOpenTextDocument(DidOpenTextDocumentParams),
    DidChangeTextDocument(DidChangeTextDocumentParams),
    DidSaveTextDocument(DidSaveTextDocumentParams),
    DidCloseTextDocument(DidCloseTextDocumentParams),
}

#[allow(clippy::large_enum_variant)]
#[derive(Debug, IntoStaticStr)]
pub(crate) enum LspRequest {
    Initialize(InitializeParams),
    DocumentFormatting(DocumentFormattingParams),
    Shutdown,
    DocumentRangeFormatting(DocumentRangeFormattingParams),
    AirViewFile(ViewFileParams),
}

#[allow(clippy::large_enum_variant)]
#[derive(Debug, IntoStaticStr)]
pub(crate) enum LspResponse {
    Initialize(InitializeResult),
    DocumentFormatting(Option<Vec<TextEdit>>),
    DocumentRangeFormatting(Option<Vec<TextEdit>>),
    Shutdown(()),
    AirViewFile(String),
}

impl std::fmt::Display for LspNotification {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        f.write_str(self.into())
    }
}
impl std::fmt::Display for LspRequest {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        f.write_str(self.into())
    }
}
impl std::fmt::Display for LspResponse {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        f.write_str(self.into())
    }
}

impl LspNotification {
    fn trace(&self) -> TraceLspNotification<'_> {
        TraceLspNotification { inner: self }
    }
}
impl LspRequest {
    fn trace(&self) -> TraceLspRequest<'_> {
        TraceLspRequest { inner: self }
    }
}
impl LspResponse {
    fn trace(&self) -> TraceLspResponse<'_> {
        TraceLspResponse { inner: self }
    }
}

struct TraceLspNotification<'a> {
    inner: &'a LspNotification,
}
struct TraceLspRequest<'a> {
    inner: &'a LspRequest,
}
struct TraceLspResponse<'a> {
    inner: &'a LspResponse,
}

impl std::fmt::Debug for TraceLspNotification<'_> {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self.inner {
            LspNotification::DidOpenTextDocument(params) => {
                // Ignore the document itself in trace logs
                f.debug_tuple(self.inner.into())
                    .field(&params.text_document.uri)
                    .field(&params.text_document.version)
                    .field(&params.text_document.language_id)
                    .finish()
            }
            _ => std::fmt::Debug::fmt(self.inner, f),
        }
    }
}

impl std::fmt::Debug for TraceLspRequest<'_> {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        std::fmt::Debug::fmt(self.inner, f)
    }
}

impl std::fmt::Debug for TraceLspResponse<'_> {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        std::fmt::Debug::fmt(self.inner, f)
    }
}

#[derive(Debug)]
struct Backend {
    /// Channel for communication with the main loop.
    events_tx: TokioUnboundedSender<Event>,

    /// Handle to main loop. Drop it to cancel the loop, all associated tasks,
    /// and drop all owned state.
    _main_loop: tokio::task::JoinSet<()>,
}

impl Backend {
    async fn request(&self, request: LspRequest) -> anyhow::Result<LspResponse> {
        tracing::info!("Incoming: {request}");
        tracing::trace!("Incoming (debug):\n{request:#?}", request = request.trace());

        let (response_tx, mut response_rx) =
            tokio_unbounded_channel::<anyhow::Result<LspResponse>>();

        // Relay request to main loop
        self.events_tx
            .send(Event::Lsp(LspMessage::Request(request, response_tx)))
            .unwrap();

        // Wait for response from main loop
        let response = response_rx.recv().await.unwrap()?;

        tracing::info!("Outgoing: {response}");
        tracing::trace!(
            "Outgoing (debug):\n{response:#?}",
            response = response.trace()
        );
        Ok(response)
    }

    fn notify(&self, notif: LspNotification) {
        tracing::info!("Incoming: {notif}");
        tracing::trace!("Incoming (debug):\n{notif:#?}", notif = notif.trace());

        // Relay notification to main loop
        self.events_tx
            .send(Event::Lsp(LspMessage::Notification(notif)))
            .unwrap();
    }

    async fn air_view_file(&self, params: ViewFileParams) -> tower_lsp::jsonrpc::Result<String> {
        cast_response!(
            self.request(LspRequest::AirViewFile(params)).await,
            LspResponse::AirViewFile
        )
    }
}

#[tower_lsp::async_trait]
impl LanguageServer for Backend {
    async fn initialize(&self, params: InitializeParams) -> Result<InitializeResult> {
        cast_response!(
            self.request(LspRequest::Initialize(params)).await,
            LspResponse::Initialize
        )
    }

    async fn initialized(&self, params: InitializedParams) {
        self.notify(LspNotification::Initialized(params));
    }

    async fn shutdown(&self) -> Result<()> {
        cast_response!(
            self.request(LspRequest::Shutdown).await,
            LspResponse::Shutdown
        )
    }

    async fn did_change_workspace_folders(&self, params: DidChangeWorkspaceFoldersParams) {
        self.notify(LspNotification::DidChangeWorkspaceFolders(params));
    }

    async fn did_change_configuration(&self, params: DidChangeConfigurationParams) {
        self.notify(LspNotification::DidChangeConfiguration(params));
    }

    async fn did_change_watched_files(&self, params: DidChangeWatchedFilesParams) {
        self.notify(LspNotification::DidChangeWatchedFiles(params));
    }

    async fn did_open(&self, params: DidOpenTextDocumentParams) {
        self.notify(LspNotification::DidOpenTextDocument(params));
    }

    async fn did_change(&self, params: DidChangeTextDocumentParams) {
        self.notify(LspNotification::DidChangeTextDocument(params));
    }

    async fn did_save(&self, params: DidSaveTextDocumentParams) {
        self.notify(LspNotification::DidSaveTextDocument(params));
    }

    async fn did_close(&self, params: DidCloseTextDocumentParams) {
        self.notify(LspNotification::DidCloseTextDocument(params));
    }

    async fn formatting(&self, params: DocumentFormattingParams) -> Result<Option<Vec<TextEdit>>> {
        cast_response!(
            self.request(LspRequest::DocumentFormatting(params)).await,
            LspResponse::DocumentFormatting
        )
    }

    async fn range_formatting(
        &self,
        params: DocumentRangeFormattingParams,
    ) -> Result<Option<Vec<TextEdit>>> {
        cast_response!(
            self.request(LspRequest::DocumentRangeFormatting(params))
                .await,
            LspResponse::DocumentRangeFormatting
        )
    }
}

pub async fn start_lsp<I, O>(read: I, write: O)
where
    I: AsyncRead + Unpin,
    O: AsyncWrite,
{
    let (service, socket) = new_lsp();
    let server = tower_lsp::Server::new(read, write, socket);
    server.serve(service).await;
}

fn new_lsp() -> (LspService<Backend>, ClientSocket) {
    let init = |client: Client| {
        let (state, events_tx) = GlobalState::new(client);

        // Start main loop and hold onto the handle that keeps it alive
        let main_loop = state.start();

        Backend {
            events_tx,
            _main_loop: main_loop,
        }
    };

    LspService::build(init)
        .custom_method("air/viewFile", Backend::air_view_file)
        .finish()
}

fn new_jsonrpc_error(message: String) -> jsonrpc::Error {
    jsonrpc::Error {
        code: jsonrpc::ErrorCode::ServerError(-1),
        message: message.into(),
        data: None,
    }
}
