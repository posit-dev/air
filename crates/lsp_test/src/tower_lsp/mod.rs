pub(crate) mod codec;

#[allow(dead_code)]
pub(crate) mod request;
