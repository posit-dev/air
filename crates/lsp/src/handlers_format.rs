//
// handlers_format.rs
//
// Copyright (C) 2024 Posit Software, PBC. All rights reserved.
//
//

use air_r_formatter::{context::RFormatOptions, format_node};
use air_r_syntax::{RExpressionList, RSyntaxKind, RSyntaxNode, WalkEvent};
use biome_formatter::{IndentStyle, LineWidth};
use biome_rowan::{AstNode, Language, SyntaxElement};
use biome_text_size::{TextRange, TextSize};
use tower_lsp::lsp_types;

use crate::state::WorldState;
use crate::{from_proto, to_proto};

#[tracing::instrument(level = "info", skip_all)]
pub(crate) fn document_formatting(
    params: lsp_types::DocumentFormattingParams,
    state: &WorldState,
) -> anyhow::Result<Option<Vec<lsp_types::TextEdit>>> {
    let doc = state.get_document(&params.text_document.uri)?;

    let line_width = LineWidth::try_from(80).map_err(|err| anyhow::anyhow!("{err}"))?;

    // TODO: Handle FormattingOptions
    let options = RFormatOptions::default()
        .with_indent_style(IndentStyle::Space)
        .with_line_width(line_width);

    if doc.parse.has_errors() {
        return Err(anyhow::anyhow!("Can't format when there are parse errors."));
    }

    let formatted = format_node(options.clone(), &doc.parse.syntax())?;
    let output = formatted.print()?.into_code();

    // Do we need to check that `doc` is indeed an R file? What about special
    // files that don't have extensions like `NAMESPACE`, do we hard-code a
    // list? What about unnamed temporary files?

    let edits = to_proto::replace_all_edit(&doc.line_index, &doc.contents, &output)?;
    Ok(Some(edits))
}

#[tracing::instrument(level = "info", skip_all)]
pub(crate) fn document_range_formatting(
    params: lsp_types::DocumentRangeFormattingParams,
    state: &WorldState,
) -> anyhow::Result<Option<Vec<lsp_types::TextEdit>>> {
    let doc = state.get_document(&params.text_document.uri)?;

    let line_width = LineWidth::try_from(80).map_err(|err| anyhow::anyhow!("{err}"))?;
    let range =
        from_proto::text_range(&doc.line_index.index, params.range, doc.line_index.encoding)?;

    // TODO: Handle FormattingOptions
    let options = RFormatOptions::default()
        .with_indent_style(IndentStyle::Space)
        .with_line_width(line_width);

    let logical_lines = find_deepest_enclosing_logical_lines(doc.parse.syntax(), range);
    if logical_lines.is_empty() {
        tracing::warn!("Can't find logical line");
        return Ok(None);
    };

    // Find the overall formatting range by concatenating the ranges of the logical lines.
    // We use the "non-whitespace-range" as that corresponds to what Biome will format.
    let format_range = logical_lines
        .iter()
        .map(text_non_whitespace_range)
        .reduce(|acc, new| acc.cover(new))
        .expect("`logical_lines` is non-empty");

    // We need to wrap in an `RRoot` otherwise the comments get attached too
    // deep in the tree. See `CommentsBuilderVisitor` in biome_formatter and the
    // `is_root` logic. Note that `node` needs to be wrapped in at least two
    // other nodes in order to fix this problem, and here we have an `RRoot` and
    // `RExpressionList` that do the job.
    //
    // Since we only format logical lines, it is fine to wrap in an expression list.
    let Some(exprs): Option<Vec<air_r_syntax::AnyRExpression>> = logical_lines
        .into_iter()
        .map(air_r_syntax::AnyRExpression::cast)
        .collect()
    else {
        tracing::warn!("Can't cast to `AnyRExpression`");
        return Ok(None);
    };

    let list = air_r_factory::r_expression_list(exprs);
    let eof = air_r_syntax::RSyntaxToken::new_detached(RSyntaxKind::EOF, "", vec![], vec![]);
    let root = air_r_factory::r_root(list, eof).build();

    let format_info = biome_formatter::format_sub_tree(
        root.syntax(),
        air_r_formatter::RFormatLanguage::new(options),
    )?;

    if format_info.range().is_none() {
        // Happens in edge cases when biome returns a `Printed::new_empty()`
        return Ok(None);
    };

    let mut format_text = format_info.into_code();

    // Remove last hard break line from our artifical expression list
    format_text.pop();
    let edits = to_proto::replace_range_edit(&doc.line_index, format_range, format_text)?;

    Ok(Some(edits))
}

// From biome_formatter
fn text_non_whitespace_range<E, L>(elem: &E) -> TextRange
where
    E: Into<SyntaxElement<L>> + Clone,
    L: Language,
{
    let elem: SyntaxElement<L> = elem.clone().into();

    let start = elem
        .leading_trivia()
        .into_iter()
        .flat_map(|trivia| trivia.pieces())
        .find_map(|piece| {
            if piece.is_whitespace() || piece.is_newline() {
                None
            } else {
                Some(piece.text_range().start())
            }
        })
        .unwrap_or_else(|| elem.text_trimmed_range().start());

    let end = elem
        .trailing_trivia()
        .into_iter()
        .flat_map(|trivia| trivia.pieces().rev())
        .find_map(|piece| {
            if piece.is_whitespace() || piece.is_newline() {
                None
            } else {
                Some(piece.text_range().end())
            }
        })
        .unwrap_or_else(|| elem.text_trimmed_range().end());

    TextRange::new(start, end)
}

/// Finds consecutive logical lines. Currently that's only expressions at
/// top-level or in a braced list.
fn find_deepest_enclosing_logical_lines(node: RSyntaxNode, range: TextRange) -> Vec<RSyntaxNode> {
    let start_lists = find_expression_lists(&node, range.start(), false);
    let end_lists = find_expression_lists(&node, range.end(), true);

    // Both vectors of lists should have a common prefix, starting from the
    // program's expression list. As soon as the lists diverge we stop.
    let Some(list) = start_lists
        .into_iter()
        .zip(end_lists)
        .take_while(|pair| pair.0 == pair.1)
        .map(|pair| pair.0)
        .last()
    else {
        // Should not happen as the range is always included in the program's expression list
        tracing::warn!("Can't find common list parent");
        return vec![];
    };

    let Some(list) = RExpressionList::cast(list) else {
        tracing::warn!("Can't cast to expression list");
        return vec![];
    };

    let iter = list.into_iter();

    // We've chosen to be liberal about user selections and always widen the
    // range to include the selection bounds. If we wanted to be conservative
    // instead, we could use this `filter()` instead of the `skip_while()` and
    // `take_while()`:
    //
    // ```rust
    // .filter(|node| range.contains_range(node.text_trimmed_range()))
    // ```
    let logical_lines: Vec<RSyntaxNode> = iter
        .map(|expr| expr.into_syntax())
        .skip_while(|node| !node.text_range().contains(range.start()))
        .take_while(|node| node.text_trimmed_range().start() <= range.end())
        .collect();

    logical_lines
}

fn find_expression_lists(node: &RSyntaxNode, offset: TextSize, end: bool) -> Vec<RSyntaxNode> {
    let mut preorder = node.preorder();
    let mut nodes: Vec<RSyntaxNode> = vec![];

    while let Some(event) = preorder.next() {
        match event {
            WalkEvent::Enter(node) => {
                let Some(parent) = node.parent() else {
                    continue;
                };

                let is_contained = if end {
                    let trimmed_node_range = node.text_trimmed_range();
                    trimmed_node_range.contains_inclusive(offset)
                } else {
                    let node_range = node.text_range();
                    node_range.contains(offset)
                };

                if !is_contained {
                    preorder.skip_subtree();
                    continue;
                }

                if parent.kind() == RSyntaxKind::R_EXPRESSION_LIST {
                    nodes.push(parent.clone());
                    continue;
                }
            }

            WalkEvent::Leave(_) => {}
        }
    }

    nodes
}

#[cfg(test)]
mod tests {
    use crate::{
        documents::Document, tower_lsp::init_test_client, tower_lsp_test_client::TestClientExt,
    };

    #[tests_macros::lsp_test]
    async fn test_format() {
        let mut client = init_test_client().await;

        #[rustfmt::skip]
        let doc = Document::doodle(
"
1
2+2
3 + 3 +
3",
        );

        let formatted = client.format_document(&doc).await;
        insta::assert_snapshot!(formatted);

        client
    }

    // https://github.com/posit-dev/air/issues/61
    #[tests_macros::lsp_test]
    async fn test_format_minimal_diff() {
        let mut client = init_test_client().await;

        #[rustfmt::skip]
        let doc = Document::doodle(
"1
2+2
3
",
        );

        let edits = client.format_document_edits(&doc).await.unwrap();
        assert!(edits.len() == 1);

        let edit = &edits[0];
        assert_eq!(edit.new_text, " + ");

        client
    }

    #[tests_macros::lsp_test]
    async fn test_format_range_none() {
        let mut client = init_test_client().await;

        #[rustfmt::skip]
        let (doc, range) = Document::doodle_and_range(
"<<>>",
        );

        let output = client.format_document_range(&doc, range).await;
        insta::assert_snapshot!(output);

        #[rustfmt::skip]
        let (doc, range) = Document::doodle_and_range(
"<<
>>",
        );

        let output = client.format_document_range(&doc, range).await;
        insta::assert_snapshot!(output);

        #[rustfmt::skip]
        let (doc, range) = Document::doodle_and_range(
"<<1
>>",
        );

        let output = client.format_document_range(&doc, range).await;
        insta::assert_snapshot!(output);

        client
    }

    #[tests_macros::lsp_test]
    async fn test_format_range_logical_lines() {
        let mut client = init_test_client().await;

        // 2+2 is the logical line to format
        #[rustfmt::skip]
        let (doc, range) = Document::doodle_and_range(
"1+1
<<2+2>>
",
        );
        let output = client.format_document_range(&doc, range).await;
        insta::assert_snapshot!(output);

        #[rustfmt::skip]
        let (doc, range) = Document::doodle_and_range(
"1+1
#
<<2+2>>
",
        );

        let output = client.format_document_range(&doc, range).await;
        insta::assert_snapshot!(output);

        // The element in the braced expression is a logical line
        // FIXME: Should this be the whole `{2+2}` instead?
        #[rustfmt::skip]
        let (doc, range) = Document::doodle_and_range(
"1+1
{<<2+2>>}
",
        );

        let output = client.format_document_range(&doc, range).await;
        insta::assert_snapshot!(output);

        #[rustfmt::skip]
        let (doc, range) = Document::doodle_and_range(
"1+1
<<{2+2}>>
",
        );
        let output = client.format_document_range(&doc, range).await;
        insta::assert_snapshot!(output);

        // The deepest element in the braced expression is our target
        #[rustfmt::skip]
        let (doc, range) = Document::doodle_and_range(
"1+1
{
  2+2
  {
    <<3+3>>
  }
}
",
        );

        let output = client.format_document_range(&doc, range).await;
        insta::assert_snapshot!(output);
        client
    }

    #[tests_macros::lsp_test]
    async fn test_format_range_mismatched_indent() {
        let mut client = init_test_client().await;

        #[rustfmt::skip]
        let (doc, range) = Document::doodle_and_range(
"1
  <<2+2>>
",
        );

        // We don't change indentation when `2+2` is formatted
        let output = client.format_document_range(&doc, range).await;
        insta::assert_snapshot!(output);

        // Debatable: Should we make an effort to remove unneeded indentation
        // when it's part of the range?
        #[rustfmt::skip]
        let (doc, range) = Document::doodle_and_range(
"1
<<  2+2>>
",
        );
        let output_wide = client.format_document_range(&doc, range).await;
        assert_eq!(output, output_wide);

        client
    }

    #[tests_macros::lsp_test]
    async fn test_format_range_multiple_lines() {
        let mut client = init_test_client().await;

        #[rustfmt::skip]
        let (doc, range) = Document::doodle_and_range(
"1+1
<<#
2+2>>
",
        );

        let output1 = client.format_document_range(&doc, range).await;
        insta::assert_snapshot!(output1);

        #[rustfmt::skip]
        let (doc, range) = Document::doodle_and_range(
"<<1+1
#
2+2>>
",
        );
        let output2 = client.format_document_range(&doc, range).await;
        insta::assert_snapshot!(output2);

        client
    }

    #[tests_macros::lsp_test]
    async fn test_format_range_unmatched_lists() {
        let mut client = init_test_client().await;

        #[rustfmt::skip]
        let (doc, range) = Document::doodle_and_range(
"0+0
<<1+1
{
  2+2>>
}
3+3
",
        );

        let output1 = client.format_document_range(&doc, range).await;
        insta::assert_snapshot!(output1);

        #[rustfmt::skip]
        let (doc, range) = Document::doodle_and_range(
"0+0
<<1+1
{
>>  2+2
}
3+3
",
        );
        let output2 = client.format_document_range(&doc, range).await;
        insta::assert_snapshot!(output2);

        #[rustfmt::skip]
        let (doc, range) = Document::doodle_and_range(
"0+0
<<1+1
{
  2+2
}
>>3+3
",
        );
        let output3 = client.format_document_range(&doc, range).await;
        insta::assert_snapshot!(output3);

        #[rustfmt::skip]
        let (doc, range) = Document::doodle_and_range(
"0+0
1+1
{
<<  2+2
}
>>3+3
",
        );
        let output4 = client.format_document_range(&doc, range).await;
        insta::assert_snapshot!(output4);

        #[rustfmt::skip]
        let (doc, range) = Document::doodle_and_range(
"<<1+1>>
2+2
",
        );

        let output5 = client.format_document_range(&doc, range).await;
        insta::assert_snapshot!(output5);

        #[rustfmt::skip]
        let (doc, range) = Document::doodle_and_range(
"1+1
<<2+2>>
",
        );

        let output6 = client.format_document_range(&doc, range).await;
        insta::assert_snapshot!(output6);

        client
    }
}
