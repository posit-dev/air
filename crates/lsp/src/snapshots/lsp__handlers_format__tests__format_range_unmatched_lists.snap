---
source: crates/lsp/src/handlers_format.rs
expression: output1
---
0+0
1 + 1
{
  2 + 2
}
3+3
